# Aprendizado Supervisionado

## Exercício 02

### Atividade 01

Finalize a construção da árvore de decisão a seguir. Utilize os conceitos de Entropia e Ganho de Informação vistos no tutorial de árvore de decisão.

<img src="http://www.data2learning.com/machinelearning/images_notebook/arvoredecisao_02.png" width="60%">


O conjunto *S* passa ser agora, no caso de chuvoso, $ S = {D4, D5, D6, D10, D14}$ que são divididos em $[3+, 2-]$.

Para finalizar a árvore, vamos trabalhar com o caso *Chuvoso*.

O primeiro passo é calcular a entropia de S ($E(S)$):

$E = \sum_{i}^{c}{-p_i\log_2{p_i}} = -\frac{3}{5}*\log_2{\frac{3}{5}} - \frac{2}{5}*\log_2{\frac{2}{5}} = 0.971$

Com o valor da Entropia de $S$, calculamos o $GI$ para cada um dos atributos restantes: Umidade, Temperatura e Vento.



### O Atributo Umidade [Alta, Normal]

O atributo umidade está dividido da seguinte forma na base de treinamento: 

* Alta: $\frac{2}{5}$ onde $[1+, 1-]$ 


* Normal: $\frac{3}{5}$ onde $[2+, 1-]$


Vamos calcular:

$- \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)} = -\frac{2}{5}*E(S_{Alta}) - \frac{3}{5}*E(S_{Normal})$

Calculando a entropia de cada atributo temos:

$E(S_{Alta}) = - \frac{1}{2}*log_2{\frac{1}{2}} - \frac{2}{2}*log_2{\frac{1}{2}} = 1$

$E(S_{Normal}) = - \frac{2}{3}*log_2{\frac{2}{3}} - \frac{1}{3}*log_2{\frac{1}{3}} = 0.918$


Agora podemos calcular o Ganho de Informação:

$- \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)} = -\frac{2}{5}*E(S_{Alta}) - \frac{3}{5}*E(S_{Normal})$

$- \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)} = -\frac{2}{5}*1 - \frac{3}{5}*0.918 = -0.951$

Por fim, temos que o GI de informação para o atributo Umidade é:

$GI(S,A) = E(S) - \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)}$

$GI(S,A) = 0.971 - 0.951 = 0.02$

### Atributo Temperatura [Quente, Intermediária, Fria]

O atributo Temperatura está dividido da seguinte forma na base de treinamento: 

* Quente: $\frac{0}{5}$ onde $[0+, 0-]$ 


* Intermediária: $\frac{3}{5}$ onde $[2+, 1-]$


* Fria: $\frac{2}{5}$ onde $[1+, 1-]$ 

É fácil definir as entropiais individiuais:

$E_(S_{Quente}) = 0$

$E_(S_{Intermediaria}) = - \frac{2}{3}*log_2{\frac{2}{3}} - \frac{1}{3}*log_2{\frac{1}{3}} = 0.918$

$E_(S_{Fria}) = 1$

Vamos calcular o Ganho de Informação:

$- \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)} = -\frac{0}{5}*E(S_{Quente}) - \frac{3}{5}*E(S_{Intermediaria}) - 
\frac{2}{5}*E(S_{Fria})$

$- \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)} = -\frac{0}{5}*0 - \frac{3}{5}0.918 - \frac{2}{5}*1 = 0.951$

Por fim, temos que o GI de informação para o atributo Temperatura é:

$GI(S,A) = E(S) - \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)}$

$GI(S,A) = 0.971 - 0.951 = 0.02$

### Atributo Vento [Fraco, Forte]

O atributo Vento está dividido da seguinte forma na base de treinamento: 

* Fraco: $\frac{3}{5}$ onde $[3+, 0-]$ 

* Forte: $\frac{2}{5}$ onde $[0+, 2-]$

É fácil definir as entropiais individiuais:

$E_(S_{Fraco}) = 0$

$E_(S_{Forte}) = 0$

Vamos calcular o Ganho de Informação:

$- \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)} = -\frac{3}{5}*E(S_{Fraco}) - \frac{2}{5}*E(S_{Forte})$

$- \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)} = -\frac{3}{5}*0 - \frac{2}{5}*0 = 0$

Por fim, temos que o GI de informação para o atributo Vento é:

$GI(S,A) = E(S) - \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)}$

$GI(S,A) = 0.971 - 0 = 0.971$

Sendo assim, temos:

$GI(S, Umidade) = 0,02$

$GI(S, Temperatura) = 0,02$

$GI(S, Vento) = 0,971$

Logo, o atributo com maior ganho de informaçõa é o `Vento`. É ele que devemos utilizar para finalizar a construção da árvore. 

A árvore final fica: 

<img src="http://www.data2learning.com/machinelearning/images_notebook/decision_tree_resposta.png" width="60%">


### Atividade 02

Considere a base de dados a seguir: 

In [2]:
import pandas

data = pandas.read_csv('http://www.data2learning.com/machinelearning/datasets/base_dt.dat')

data

Idade Salario SuperiorCompleto Dependentes Atrasou
0   MENOR30    ALTO              NAO         NAO     NAO
1   MENOR30    ALTO              NAO         SIM     NAO
2      3140    ALTO              NAO         NAO     SIM
3   MAIOR40   MEDIO              NAO         NAO     SIM
4   MAIOR40   BAIXO              SIM         NAO     SIM
5   MAIOR40   BAIXO              SIM         SIM     NAO
6      3140   BAIXO              SIM         SIM     SIM
7   MENOR30   MEDIO              NAO         NAO     NAO
8   MENOR30   BAIXO              SIM         SIM     SIM
9   MAIOR40   MEDIO              SIM         NAO     SIM
10  MENOR30   MEDIO              SIM         SIM     SIM
11     3140   MEDIO              NAO         SIM     SIM
12     3140    ALTO              SIM         NAO     SIM
13  MAIOR40   MEDIO              NAO         SIM     NAO

In [3]:
from sklearn import preprocessing

le_P = preprocessing.LabelEncoder() # Panorama
le_T = preprocessing.LabelEncoder() # Temperatura
le_U = preprocessing.LabelEncoder() # Umidade
le_V = preprocessing.LabelEncoder() # Vento
le_J = preprocessing.LabelEncoder() # Jogar

# Transforma os atributos de string em numéricos. Os números são atribuídos de acordo com a posição do label 
# na lista que está sendo impressa.
data.Idade = le_P.fit_transform(data.Idade)
print("Valores para o atributo Idade: ", list(le_P.classes_))

data.Salario = le_T.fit_transform(data.Salario)
print("Valores para o atributo Salário: ", list(le_T.classes_))

data.SuperiorCompleto = le_U.fit_transform(data.SuperiorCompleto)
print("Valores para o atributo SuperiorCompleto: ", list(le_U.classes_))

data.Dependentes = le_V.fit_transform(data.Dependentes)
print("Valores para o atributo Dependentes: ", list(le_V.classes_))

data.Atrasou = le_J.fit_transform(data.Atrasou)
print("Valores para o atributo Atrasou: ", list(le_J.classes_))

Valores para o atributo Idade:  ['3140', 'MAIOR40', 'MENOR30']
Valores para o atributo Salário:  [' ALTO', ' BAIXO', ' MEDIO']
Valores para o atributo SuperiorCompleto:  [' NAO', ' SIM']
Valores para o atributo Dependentes:  [' NAO', ' SIM']
Valores para o atributo Atrasou:  [' NAO', ' SIM']


A base analisa algumas informações sobre clientes de um plano de saúde e analisa se eles atrasaram ou não o pagamento da mensalidade do plano. Deseja-se construir um modelo de predição baseado em árvore de decisão. O primeiro passo é determinar qual o atributo raiz da árvore de decisão. Calcule qual o melhor atributo para a raiz da árvore sabendo que que o Ganho de Informação (GI) do atributo Salário é 0.03 e o do atributo SuperiorCompleto é 0.152.

Sabendo que o cálculo do GI é dado por: 

$GI(S,A) = E(S) - \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)}$, 

onde S é o conjunto de treinamento que vai ser dividido e A é o atributo a ser considerado. 

O primeiro passo é calcular E(S).

$E(S) = -\frac{9}{14}\log_2{\frac{9}{14}} - \frac{5}{14}\log_2{\frac{5}{14}} = 0.940$


Agora vamos calcular a entropia em cada um dos atributos restantes para que possamos analisar o ganho de informação.

### `Atributo: Dependentes`

O primeiro atributo a ser analisado será **Dependentes** (que pode receber dois valores: Sim ou Não). Como visto, $S$ possui 9 instâncias positivas e 5 instâncias negativas (classe para classificação). Considere também que 4 dos exemplos positivos e 3 dos exemplos negativos são associados a ter dependentes. Além disso, 2 exemplos negativos e 5 exemplos positivos estão associados a não ter dependentes.

Vamos calcular o ganho de informação ao selecionar o  atributo Dependentes para a raiz de uma árvore de decisão. Vejamos como é feito o cálculo, utilizando as definições matemáticas apresentadas no material didático.

**Cálculo**

O conjunto S está dividido da seguinte forma: $S = [9+, 5-]$ e a divisão de acordo com o atributo `Dependentes` é dada por:

$S_{Sim} \leftarrow [4+, 3-]$

$S_{Não} \leftarrow [5+, 2-]$

Como visto, o atributo Dependentes possui dois valores e devemos calcular a entropia para cada um desses atributos: 

$E(S_{sim}) = -\frac{4}{7}\log_2{\frac{4}{7}} - \frac{3}{7}\log_2{\frac{3}{7}} = 0.985$

$E(S_{nao}) = -\frac{5}{7}\log_2{\frac{5}{7}} - \frac{2}{7}\log_2{\frac{2}{7}} = 0.863$

Com isso podemos calcular: 

$- \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)}$

$- \frac{7}{14}*0.985 - \frac{7}{14}*0.863 = -0.924$

Por fim, temos que  o Ganho de Informação é: 

$GI(S,A) = E(S) - \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)} = 0.940 - 0.924 = 0.016$

### `Atributo: Idade`

Vamos trabalhar da mesma forma que fizemos o atributo anterior. O conjunto S está dividido da seguinte forma: $S = [9+, 5-]$ e a divisão de acordo com o atributo `Idade` é dada por:

$S_{MENOR30} \leftarrow [2+, 3-]$

$S_{3140} \leftarrow [4+, 0-]$

$S_{MAIOR40} \leftarrow [3+, 2-]$

Como visto, o atributo Idade possui três valore e devemos calcular a entropia para cada um desses atributos: 

$E(S_{MENOR30}) = -\frac{2}{5}\log_2{\frac{2}{5}} - \frac{3}{5}\log_2{\frac{3}{5}} = 0.971$

$E(S_{3140}) = 0$

$E(S_{MAIOR40}) = -\frac{3}{5}\log_2{\frac{3}{5}} - \frac{2}{5}\log_2{\frac{2}{5}} = 0.971$

Com isso podemos calcular: 

$- \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)}$

$- \frac{5}{14}*0.971 - \frac{4}{14}*0 - \frac{5}{14}*0.971 = -0.693$

Por fim, temos que  o Ganho de Informação é: 

$GI(S,A) = E(S) - \sum_{v \in Valores(A)}{\frac{S_v}{S}E(S_v)} = 0.940 - 0.693 = 0.247$

Sendo assim, temos:

* Atributo Idade: GI = 0.247
* Atributo Salário: GI = 0.03
* Atributo SuperiorCompleto: GI = 0.152
* Atributo Dependentes: GI = 0.016

Logo, o melhor atributo para ser raiz da árvore é o atributo `Idade`.

### Atividade 03

Para esta ativididade vamos trabalhar com a base de dígitos disponibilizada pelo scikit-learn.

Detalhes da base podem ser econtrados [neste link](http://archive.ics.uci.edu/ml/datasets/Pen-Based+Recognition+of+Handwritten+Digits).

A tarefa é classificar um conjunto de imagens de acordo com o dígito que a imagem corresponde. As imagens foram coletadas a partir de dígitos escritos à mão. 

Tarefa:

* Aplique o algoritmo de árvore de decisão e reporte o valor da acurácia 
* Aplique o algoritmo KNN e compare o modelo de KNN (com o melhor valor de k) com o modelo de árvore de decisão (variando também os valores necessários).
* Em todos os testes utilize a validação cruzada com 5 folds.

In [6]:
from sklearn import datasets
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score

digits = datasets.load_digits()

X = digits.data
y = digits.target

### Aplicando KNN para escolha do melhor valor de `k`

In [10]:
range_k = range(1, 30)

best_acc = 0
best_k = 0

for k in range_k:
    knn_ = KNeighborsClassifier(n_neighbors=k)
    
    score_ = cross_val_score(knn_, X, y, cv=5, scoring='accuracy')
    
    score_mean = score_.mean()
    
    if score_mean > best_acc:
        best_k = k
        best_acc = score_mean
        
    print(k, score_mean)

print("Best: %i: %f" % (best_k, best_acc))

1 0.9649752094799343
2 0.9666281215002192
3 0.9661143537042125
4 0.9639057779662252
5 0.9627899114966898
6 0.9594189778290522
7 0.9600006636118843
8 0.9577720868648678
9 0.9561253164543428
10 0.9550157272024308
11 0.9555621421134866
12 0.9561099260379361
13 0.9555511626324675
14 0.9560991524977508
15 0.9544338354787648
16 0.9544509214030586
17 0.9533120120025
18 0.948874859517496
19 0.9516358008896691
20 0.9505368144035937
21 0.948882476865627
22 0.9499982579479882
23 0.9483299906777816
24 0.949461247563724
25 0.9461061952977359
26 0.945564517816561
27 0.9449844572086453
28 0.9427435256836165
29 0.9405460148138225
Best: 2: 0.966628


O melhor modelo usando KNN é quando `k` é igual a `2` e possui uma acurácia de `0.966628`

### Árvore de Decisão

Vamos variar alguns parâmetros da árvore de decisão para tentar encontrar um modelo melhor que o primeiro encontrado. A lista completa de atributos está disponível [neste link](http://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html).

O primeiro teste da árvore de decisão será utilizando os valores padrões. 

In [13]:
tree_clf = DecisionTreeClassifier()
tree_clf

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [14]:
score_ = cross_val_score(tree_clf, X, y, cv=5, scoring='accuracy')
print("Árvore de Decisão - Acc: %f" % score_.mean())

Árvore de Decisão - Acc: 0.769245


Vamos variar um dos parâmetros: 

```
criterion : string, optional (default=”gini”)

    The function to measure the quality of a split. Supported criteria are “gini” for the Gini impurity and “entropy” for the information gain.
```

In [15]:
tree_clf = DecisionTreeClassifier(criterion='entropy')
score_ = cross_val_score(tree_clf, X, y, cv=5, scoring='accuracy')
print("Árvore de Decisão - Acc: %f" % score_.mean())

Árvore de Decisão - Acc: 0.801313


O `information gain` como critério para separação dos nós gerou um modelo melhor do que o que utilizava `Gini Impurity`. Vamos usar esse atributo nos próximos exemplos. 

In [31]:
tree_clf = DecisionTreeClassifier(criterion='entropy', max_depth=22)
score_ = cross_val_score(tree_clf, X, y, cv=5, scoring='accuracy')
print("Árvore de Decisão - Acc: %f" % score_.mean())

Árvore de Decisão - Acc: 0.810796


In [44]:
import numpy as np 

best_acc = 0
best_depth = 0

max_depths = np.linspace(1, 50, 50, endpoint=True)
print("Decision Tree: ")
for max_depth in max_depths:
    tree_clf = DecisionTreeClassifier(criterion='entropy', max_depth=max_depth, random_state=int(max_depth))
    score_ = cross_val_score(tree_clf, X, y, cv=5, scoring='accuracy')
    
    if score_.mean() > best_acc:
        best_acc = score_.mean()
        best_depth = max_depth
    
    print("Depth %i - Acc: %f" % (max_depth, score_.mean()))
    
print("Best: %i: %f" % (best_depth, best_acc))

Decision Tree: 
Depth 1 - Acc: 0.195339
Depth 2 - Acc: 0.328135
Depth 3 - Acc: 0.502474
Depth 4 - Acc: 0.632277
Depth 5 - Acc: 0.734772
Depth 6 - Acc: 0.772502
Depth 7 - Acc: 0.787070
Depth 8 - Acc: 0.804730
Depth 9 - Acc: 0.810308
Depth 10 - Acc: 0.808056
Depth 11 - Acc: 0.807547
Depth 12 - Acc: 0.806413
Depth 13 - Acc: 0.809741
Depth 14 - Acc: 0.799244
Depth 15 - Acc: 0.799716
Depth 16 - Acc: 0.805803
Depth 17 - Acc: 0.804124
Depth 18 - Acc: 0.795367
Depth 19 - Acc: 0.804731
Depth 20 - Acc: 0.806829
Depth 21 - Acc: 0.804756
Depth 22 - Acc: 0.800163
Depth 23 - Acc: 0.806922
Depth 24 - Acc: 0.805299
Depth 25 - Acc: 0.801434
Depth 26 - Acc: 0.799165
Depth 27 - Acc: 0.805904
Depth 28 - Acc: 0.807603
Depth 29 - Acc: 0.805223
Depth 30 - Acc: 0.813089
Depth 31 - Acc: 0.799210
Depth 32 - Acc: 0.802494
Depth 33 - Acc: 0.806405
Depth 34 - Acc: 0.801458
Depth 35 - Acc: 0.802454
Depth 36 - Acc: 0.804759
Depth 37 - Acc: 0.795249
Depth 38 - Acc: 0.801347
Depth 39 - Acc: 0.814702
Depth 40 - Acc: 0.

Como foi visto no material de Árvore de Decisão a poda é um importante parâmetro que ajuda evitar `overfitting`. Vamos usar profundidade `39` que nos retornou uma acurácia de `0.814702`. 

Tentamos variar `min_samples_splits` e `min_samples_leaf` mas não teve melhoria expressiva nos modelos.